databricks/koalas/groupby.py

```diff
@@ -563,8 +563,11 @@ def diff(self, periods=1):
         5    NaN
         Name: a, dtype: float64
         """
-        return self._apply_series_op(
-            lambda sg: sg._kser._diff(periods, part_cols=sg._groupkeys_scols))
+        if self._kdf.index.is_monotonic or self._kdf.index.is_monotonic_decreasing:
```

Comment:
Shall we use the similar logic to what we did in `loc` indexer?
This can invoke Spark job twice.

Reply:
Btw, why do you know are these related to index ordering?

Reply:
Seems like it's not related to index ordering?

```py
>>> pd.__version__
>>> df = pd.DataFrame({'a': [1,2,3,4,5,6], 'b': [1,1,2,3,5,8], 'c': [1,4,9,16,25,36]}, index=[10,0,20,30,5,15])
'1.0.0'
>>> df
    a  b   c
10  1  1   1
0   2  1   4
20  3  2   9
30  4  3  16
5   5  5  25
15  6  8  36
>>> df.groupby(['b']).diff().sort_index()
      a    c
0   1.0  3.0
5   NaN  NaN
10  NaN  NaN
15  NaN  NaN
20  NaN  NaN
30  NaN  NaN
```

If you saw such a behavior, it could be the different reason.
Could you share the code we can reproduce?

Reply:
Ahh, I found the reproducible example:

```py
>>> df = pd.DataFrame({'a': [1, 2, 3, 4, 5, 6], 'b': [1, 1, 2, 3, 5, 8], 'c': [1, 4, 9, 16, 25, 36]}, index=np.random.rand(6))
>>> df.groupby(['b']).diff().sort_index()
Traceback (most recent call last):
...
```

Actually I suspect it's a pandas' bug.

Reply:
thanks for the researching.

and i also tested several examples like the below,

```python
>>> df = pd.DataFrame({'a': [1,2,3,4,5,6], 'b': [1,1,2,3,5,8], 'c': [1,4,9,16,25,36]}, index=[0.4, 0.9, 0.0, 0.6, 0.5, 0.1])
>>> df.groupby(['b']).diff()
       a    c
0.4  NaN  NaN
0.9  1.0  3.0
0.0  NaN  NaN
0.6  NaN  NaN
0.5  NaN  NaN
0.1  NaN  NaN
>>> df = pd.DataFrame({'a': [1,2,3,4,5,6], 'b': [1,1,2,3,5,8], 'c': [1,4,9,16,25,36]}, index=[0.46, 0.97, 0.00, 0.69, 0.55, 0.15])
>>> df.groupby(['b']).diff()
        a    c
0.46  NaN  NaN
0.97  1.0  3.0
0.00  NaN  NaN
0.69  NaN  NaN
0.55  NaN  NaN
0.15  NaN  NaN
>>> df = pd.DataFrame({'a': [1,2,3,4,5,6], 'b': [1,1,2,3,5,8], 'c': [1,4,9,16,25,36]}, index=[0.465, 0.970, 0.004, 0.690, 0.551, 0.155])
>>> df.groupby(['b']).diff()
Traceback (most recent call last):
...
```

so we can guess it might be pandas' bug when index is not monotonic and when float-type index has more than 3 decimal points.

Reply:
oops.. but the below pandas example doesn't raise exception even the float-type index has more than 3 decimal points.

```python
>>> df = pd.DataFrame({'a': [1,2,3,4,5,6], 'b': [1,1,2,3,5,8], 'c': [1,4,9,16,25,36]}, index=[0.46101, 0.9700, 0.0000, 0.6900, 0.5500, 0.1500])
>>> df.groupby(['b']).diff()
           a    c
0.46101  NaN  NaN
0.97000  1.0  3.0
0.00000  NaN  NaN
0.69000  NaN  NaN
0.55000  NaN  NaN
0.15000  NaN  NaN
```

hmm.... i think it's totally not reasonable.

Reply:
I'd say we can just ignore this behavior for now.
I mean, just revert the changes around here.
WDYT? cc @HyukjinKwon

Reply:
> I mean, just revert the changes around here.

@ueshin yeah i totally agree with that since this implementation seems to be too overkilling.

Reply:
then i'm keeping this for now until @HyukjinKwon response about this.

Reply:
+1 for https://github.com/databricks/koalas/pull/1197/files#r376736123